src/GenerateRoutes.zig

```diff
@@ -0,0 +1,272 @@
+
+        const extension = std.fs.path.extension(entry.path);
+        const basename = std.fs.path.basename(entry.path);
+
+        if (std.mem.eql(u8, basename, "routes.zig")) continue;
+        if (std.mem.eql(u8, basename, "zmpl.manifest.zig")) continue;
+        if (std.mem.startsWith(u8, basename, ".")) continue;
+        if (!std.mem.eql(u8, extension, ".zig")) continue;
+
+        const routes = try self.generateRoute(views_dir, entry.path);
+
+        for (routes.static) |route| {
+            try self.static_routes.append(route);
+        }
+
+        for (routes.dynamic) |route| {
+            try self.dynamic_routes.append(route);
+        }
+    }
+
+    std.sort.pdq(Function, self.static_routes.items, {}, Function.lessThanFn);
+    std.sort.pdq(Function, self.dynamic_routes.items, {}, Function.lessThanFn);
+
+    try writer.writeAll("pub const routes = struct {\n");
+    try writer.writeAll("    pub const static = .{\n");
+
+    for (self.static_routes.items) |static_route| {
+        try self.writeRoute(writer, static_route);
+    }
+
+    try writer.writeAll("    };\n\n");
+    try writer.writeAll("    pub const dynamic = .{\n");
+
+    for (self.dynamic_routes.items) |dynamic_route| {
+        try self.writeRoute(writer, dynamic_route);
+        const name = try dynamic_route.fullName(self.allocator);
+        defer self.allocator.free(name);
+        std.debug.print("[jetzig] Imported route: {s}\n", .{name});
+    }
+
+    try writer.writeAll("    };\n");
+    try writer.writeAll("};");
+}
+
+fn writeRoute(self: *Self, writer: std.ArrayList(u8).Writer, route: Function) !void {
+    const full_name = try route.fullName(self.allocator);
+    defer self.allocator.free(full_name);
+
+    const uri_path = try route.uriPath(self.allocator);
+    defer self.allocator.free(uri_path);
+
+    const output_template =
+        \\        .{{
+        \\            .name = "{s}",
+        \\            .action = "{s}",
+        \\            .uri_path = "{s}",
+        \\            .template = "{s}",
+        \\            .function = @import("{s}").{s},
+        \\        }},
+        \\
+    ;
+
+    const output = try std.fmt.allocPrint(self.allocator, output_template, .{
+        full_name,
+        route.name,
+        uri_path,
+        full_name,
+        route.path,
+        route.name,
+    });
+
+    defer self.allocator.free(output);
+    try writer.writeAll(output);
+}
+
+const RouteSet = struct {
+    dynamic: []Function,
+    static: []Function,
+};
+
+fn generateRoute(self: *Self, views_dir: std.fs.Dir, path: []const u8) !RouteSet {
+    // REVIEW: Choose a sensible upper limit or allow user to take their own risks here ?
+    const stat = try views_dir.statFile(path);
+    const source_buf = try views_dir.readFileAlloc(self.allocator, path, stat.size);
+    defer self.allocator.free(source_buf);
+
+    const source = try std.mem.concatWithSentinel(self.allocator, u8, &[_][]const u8{source_buf}, 0);
```

Comment:
Instead of doing this concat, use readFileAllocOptions in the views_dir

Reply:
GREAT - thanks, this felt so clunky.